### PR TITLE
chore: Unify the builder methods

### PR DESCRIPTION
**Pull request by FrankEssenberger**
In this PR the builder methods are adjusted so that they behave in a unified way.
Have a look at the ADR 0017 for a description of the API decisions.

### CODE REVIEW

File: knowledge-base/adr/0017-code-snippets.ts [hunk 1]
```diff
@@ -0,0 +1,127 @@
+/* eslint-disable */
+namespace Classes{
```
Comment (marikaner) [inline knowledge-base/adr/0017-code-snippets.ts:2]: Not sure if this really needs to be here?

File: packages/core/src/odata-common/request-builder/get-request-builder-base.ts [hunk 2]
```diff
@@ -0,0 +1,50 @@
+import { variadicArgumentToArray } from '@sap-cloud-sdk/util';
+import { Constructable, EntityIdentifiable, EntityBase } from '../entity';
+import { Selectable } from '../selectable/selectable';
+import { ODataGetAllRequestConfig } from '../request/odata-get-all-request-config';
+import { MethodRequestBuilderBase } from '../request-builder/request-builder-base';
+import { ODataGetByKeyRequestConfig } from '../request';
+
+/**
+ * Base class for the get all and get by key request builders
```
Comment (marikaner) [inline packages/core/src/odata-common/request-builder/get-request-builder-base.ts:9]:
```suggestion
 * Base class for the get all and get by key request builders.
```

File: packages/core/src/odata-common/request-builder/get-request-builder-base.ts [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+ *
+ * @typeparam EntityT - Type of the entity to be requested
```
Comment (marikaner) [inline packages/core/src/odata-common/request-builder/get-request-builder-base.ts:11]:
```suggestion
 * @typeparam EntityT - Type of the entity to be requested.
```

File: packages/core/src/odata-common/request-builder/get-request-builder-base.ts [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+      | ODataGetByKeyRequestConfig<EntityT>
+  >
+  extends MethodRequestBuilderBase<RequestConfigT>
+  implements EntityIdentifiable<EntityT> {
+  readonly _entity: EntityT;
+
+  /**
+   * Creates an instance of GetAllRequestBuilder.
+   *
+   * @param _entityConstructor - Constructor of the entity to create the request for
```
Comment (marikaner) [inline packages/core/src/odata-common/request-builder/get-request-builder-base.ts:26]:
```suggestion
   * @param _entityConstructor - Constructor of the entity to create the request for.
```

File: packages/core/src/odata-common/request-builder/get-request-builder-base.ts [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    readonly _entityConstructor: Constructable<EntityT>,
+    requestConfig: RequestConfigT
+  ) {
+    super(requestConfig);
+  }
+  /**
+   * Restrict the response to the given selection of properties in the request.
+   *
+   * @param selects - Fields to select in the request
+   * @returns The request builder itself, to facilitate method chaining
```
Comment (marikaner) [inline packages/core/src/odata-common/request-builder/get-request-builder-base.ts:39]:
```suggestion
   * @param selects - Fields to select in the request.
   * @returns The request builder itself, to facilitate method chaining.
```

File: packages/core/src/odata-common/request-builder/request-builder-base.ts [hunk 6]
```diff
@@ -53,35 +53,72 @@ export abstract class MethodRequestBuilderBase<
...
    * @returns The request builder itself, to facilitate method chaining
+   * @deprecated Since version 1.34.0 Use [[addCustomHeaders]] instead.
    */
   withCustomHeaders(headers: Record<string, string>): this {
     this.requestConfig.addCustomHeaders(headers);
     return this;
   }
 
+  /**
+   * Add custom headers to the request. If a header field with the given name already exists it is overwritten.
```
Comment (marikaner) [inline packages/core/src/odata-common/request-builder/request-builder-base.ts:64]: Not sure, but somehow I stumble over this sentence.
```suggestion
   * Add custom headers to the request. Existing headers will be overwritten.
```

File: packages/core/src/odata-common/request-builder/request-builder-base.ts [hunk 7]
```diff
@@ -53,35 +53,72 @@ export abstract class MethodRequestBuilderBase<
...
+   *
+   * @param headers - Key-value pairs denoting additional custom headers
+   * @returns The request builder itself, to facilitate method chaining
```
Comment (marikaner) [inline packages/core/src/odata-common/request-builder/request-builder-base.ts:67]:
```suggestion
   * @param headers - Key-value pairs denoting additional custom headers.
   * @returns The request builder itself, to facilitate method chaining.
```

**Review by marikaner — APPROVED** [review]
LGTM, I started adding points in the ends of every Typedoc comment (also for params). But I am not sure if this is / should be a rule. You can also ignore it ;)

File: knowledge-base/adr/0017-code-snippets.ts (hunk 1 above)
Comment (FrankEssenberger) [reply knowledge-base/adr/0017-code-snippets.ts:2]: I would like to keep it as an explanation what has been tried and explain the options.

**Pull request merged**